### PR TITLE
Goldilocks inverse

### PR DESCRIPTION
**Pull request by 0xKanekiKen**
This PR adds method to compute inverse of a field element in Goldilocks in constant time.

### CODE REVIEW

File: goldilocks/src/lib.rs [hunk 1]
```diff
@@ -135,8 +135,57 @@ impl Field for Goldilocks {
...
+        // compute base^111111111111 (12 ones)
+        // repeatedly square t6 6 times and multiply by t6
+        let t12 = exp_acc::<6>(t6, t6);
+
+        // compute base^111111111111111111111111 (24 ones)
+        // repeatedly square t12 12 times and multiply by t12
+        let t24 = exp_acc::<12>(t12, t12);
+
+        // compute base^1111111111111111111111111111111 (31 ones)
+        // repeatedly square t24 7 times and multiply by t6
```
Comment (npwardberkeley) [inline goldilocks/src/lib.rs:179]: Isn't this multiplying by `t6^2*base`, not by `t6`?

File: goldilocks/src/lib.rs [hunk 2]
```diff
@@ -135,8 +135,57 @@ impl Field for Goldilocks {
...
+        let t30 = exp_acc::<6>(t24, t6);
+        let t31 = t30.square() * *self;
+
+        // compute base^111111111111111111111111111111101111111111111111111111111111111
+        // repeatedly square t31 31 times and multiply by t31
```
Comment (npwardberkeley) [inline goldilocks/src/lib.rs:184]: 32 times?

**Review by npwardberkeley — COMMENTED** [review]
Nice work!! Looks good overall, a few nits about the comments

File: goldilocks/src/lib.rs (hunk 2 above)
Comment (0xKanekiKen) [reply goldilocks/src/lib.rs]: yes- fixed it in [1192f34](https://github.com/Plonky3/Plonky3/pull/40/commits/1192f34f4aca1d7b693f46c09c4fb8f7ce60e897)

File: goldilocks/src/lib.rs (hunk 1 above)
Comment (0xKanekiKen) [reply goldilocks/src/lib.rs]: the comment is confusing and I've updated it to the following: 

`compute base^1111111111111111111111111111111 (31 ones)`
`repeatedly square t24 6 times and multiply by t6 first. then square t30 and multiply by base`

I hope this is fine

File: goldilocks/src/lib.rs [hunk 3]
```diff
@@ -262,6 +312,19 @@ impl Div for Goldilocks {
...
+// ================================================================================================
+
+/// Squares the base N number of times and multiplies the result by the tail value.
+#[inline(always)]
+fn exp_acc<const N: usize>(base: Goldilocks, tail: Goldilocks) -> Goldilocks {
+    let mut result = base;
+    for _ in 0..N {
+        result = result.square();
+    }
+    result * tail
```
Comment (dlubarov) [inline goldilocks/src/lib.rs:325]: nit: there's an `exp_power_of_2` which almost does this

```suggestion
    base.exp_power_of_2(N) * tail
```

**Review by dlubarov — APPROVED** [review]
We also had what seemed like a pretty optimized implementation in Plonky2 (by @unzvfu). At first glance I'd expect this approach to be slower, but I think it's fine to merge this for now and we can benchmark later.

File: goldilocks/src/lib.rs (hunk 3 above)
Comment (0xKanekiKen) [reply goldilocks/src/lib.rs]: thanks for the suggestion.

**Comment by 0xKanekiKen** [comment]
Thanks, @dlubarov, for mentioning the other implementation. I'd be happy to do the benchmark analysis of the optimized implementation from Plonky2. For now, I've added a TODO for the benchmark.

**Comment by unzvfu** [comment]
FWIW, I would be also be surprised if this approach is faster than the version in https://github.com/mir-protocol/plonky2/blob/main/field/src/inversion.rs since this version uses the field's modular multiplication whereas the existing one avoids them entirely. I would vote to **benchmark first** and merge only once Daniel and my intuition has been proved wrong. (Which, to be clear, would be a pleasant and welcome surprise!)

**Comment by 0xKanekiKen** [comment]
Sure, I'd be happy to do a benchmark analysis of both these versions and get back to you.

**Comment by JL2718** [comment]
Another baseline here: #44 
vanilla recursive extended euclidean in i128

**Comment by 0xKanekiKen** [comment]
I run the benchmark on `AMD Ryzen 7 PRO 6850U` running `MANJARO` and I'm seeing the following numbers:

`try_inverse`                                         time:   **[140.87 ns 141.16 ns 141.58 ns]**

I'll check how to incorporate the optimized version into plonky3 and share it's number here.

**Comment by unzvfu** [comment]
Ok, I can confirm that I get exactly the same numbers (~142ns) for your implementation on my machine (Intel Core i9-11950H @ 2.60GHz). And indeed to my surprise the benchmark for my inverse implementation in Plonky2 gives

> `try_inverse` time:   **[164.46 ns 164.62 ns 164.91 ns]**

So that's an improvement of about 15%. Nicely done! :+1: 

It will be interesting to see which of the two methods is faster for the smaller field sizes. My GCD-based version suffers somewhat from needing to use `i128`s sometimes when the inputs are `u64`s, but for the smaller field sizes everything will be in `i64`, so it might make up some ground there. We'll see.

Anyway, this is a clear improvement so I think this is good to merge.

**Comment by 0xKanekiKen** [comment]
Awesome, I'm glad to hear that we're seeing some performance gains! Thank you, @unzvfu, for confirming and sharing your thoughts on why the introduced implementation might outperform the `plonky2` one. Your explanation about the use of `i128s` vs `i64s` makes total sense.

Just to clarify, when you mentioned smaller field sizes, were you referring to `baby bear` and `Mersenne` prime fields? I'd be happy to run a benchmark on these two small fields.

**Comment by unzvfu** [comment]
> Just to clarify, when you mentioned smaller field sizes, were you referring to `baby bear` and `Mersenne` prime fields? I'd be happy to run a benchmark on these two small fields.

Yes, exactly. It would be good for the inverse benchmarking to support those fields too (similar to how we did it in Plonky2), though there's no rush. (Also, my implementation of modular inverse will need some tweaking before a fair comparison can take place. I'll probably do that eventually, but it's not a priority at the moment.)

For now, I think there are some clippy warnings that need to be addressed, and we can merge once you've pushed a fix for those.

**Comment by JL2718** [comment]
In #44 I have a GCD inverse which does not require i128 in Goldilocks.

It's implemented in p3_field::PrimeField64/32 as a default so this implementation will just override it.

**Comment by Nashtare** [comment]
@unzvfu given the nice improvement, should we consider backporting this to plonky2 as well?

**Comment by unzvfu** [comment]
> @unzvfu given the nice improvement, should we consider backporting this to plonky2 as well?

Sure! I'm happy to review the PR if you're happy to prepare it. :)

**Pull request merged**